Add BPF 101 README file

jakule opened the pull request:
Added a README document to introduce and guide new users on BPF (Berkeley Packet Filter) code writing. The file includes useful links, described the environment setup and outlined the supported architectures. This will provide the necessary resources and steps on how to use and develop BPF codes.

jakule committed:
Add BPF 101 README file

Added a README document to introduce and guide new users on BPF (Berkeley Packet Filter) code writing. The file includes useful links, described the environment setup and outlined the supported architectures. This will provide the necessary resources and steps on how to use and develop BPF codes.

jakule committed:
Improve README with BPF code structure and usage details

Extended the README document in the BPF directory to include a comprehensive explanation on the structure of BPF code and its usage aspects. This elaboration covers the distinction between BPF program and user space code, their locations in the repository, certain limitations, their compilation process, as well as compatibility issues concerning the `libbpfgo` library. This addition is expected to provide clearer guidance for developers working with BPF code in this project.

jakule committed:
Add BPF helpers documentation link and license details to README

Added a link to the BPF helpers documentation in the README for easy reference. Also included detailed information about the license requirement for BPF programs in the kernel. This is aimed to inform developers about the GPL default setting for the license field in ELF files, and the limitations imposed by using the `Dual BSD/GPL` license in Teleport. Instructions for enabling all BPF features have also been provided.

jakule committed:
Add logging section to BPF README.md

Included a new section in the BPF README.md file detailing how to enable logging within BPF programs. This change provides clear instructions and a useful resource link to developers on how to set the license to 'GPL' for BPF programs, and how to use the 'bpf_printk' helper for logging messages to the kernel log. Adding this documentation promotes a better understanding of BPF logging functions and their license requirements.

jakule committed:
Enhance BPF README with BPF-user space communication info

Expanded the BPF README.md file to include a section about communication between BPF programs and user space via maps. This addition provides a valuable explanation and example of how both user space and BPF programs can access maps. This highlights the interaction between BPF and user space, educating developers on properly utilizing BPF-helper for map access in BPF programs and 'libbpfgo' library in user space.

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+
+#### Before you start here are some useful links:
+
+https://docs.kernel.org/bpf/ - BPF documentation
+https://github.com/aquasecurity/libbpfgo - Go bindings for libbpf
+https://github.com/iovisor/bcc - BPF Compiler Collection (a lot of examples)
+https://github.com/cilium/cilium - BPF-based networking, security, and observability (more examples)
+https://man7.org/linux/man-pages/man7/bpf-helpers.7.html - BPF helpers documentation
```
```suggestion
* https://docs.kernel.org/bpf/ - BPF documentation
* https://github.com/aquasecurity/libbpfgo - Go bindings for libbpf
* https://github.com/iovisor/bcc - BPF Compiler Collection (a lot of examples)
* https://github.com/cilium/cilium - BPF-based networking, security, and observability (more examples)
* https://man7.org/linux/man-pages/man7/bpf-helpers.7.html - BPF helpers documentation
```

Markdown joins the lines otherwise.

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+We support BPFs only on ARM64 and x86_64 architectures.
+
+1. VM with Ubuntu 20.04+ (or any other Linux distro with the kernel 5.8+ - Docker doesn't work for many BPF features)
+2. Install `clang` and `llvm` (required for compiling BPF code)
```
Does `llvm` needs to be explicitly installed, or is whatever clang pulls good enough?

jakule replied:
`clang` should be enough. I'll fix that.

codingllama commented on bpf/README.md:
````diff
@@ -0,0 +1,69 @@
+You can also use our Dockerfile to build the environment:
+
+```bash
+make -C build.assets release-centos7
````
Didn't you say "no Centos7"?

jakule replied:
BPF doesn't run on CentOS 7, as it shipped with kernel 3.10, but our release build uses CentOS 7, and that is the build that we the most take care of.
Build on CentOS 7, run on Ubuntu. I think that the easiest way to put it.

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+
+BPF code can be divided into two parts:
+
+1. BPF program - the actual BPF code that will be loaded into the kernel
```
```suggestion
1. BPF program - the actual BPF code that will be executed by the kernel
```

Disregard if not accurate.

jakule replied:
I think they are both correct. The program is first loaded and then executed, but I'd argue that loaded is more accurate. You load the program into the kernel, but the execution happens when there is an event that triggers it (you always load, execution technically may never happen).

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+The BPF programs are compiled using `clang` 10+, GCC is not supported. 
+
+User space code is located in `lib/bpf` directory. It's written in Go and uses `libbpfgo` library to load BPF programs into the kernel.
+`libbpfgo` is a Go wrapper around `libbpf` library. It's a low-level library that allows you to load BPF programs into the kernel.
```
```suggestion
`aquasecurity/libbpfgo` is a Go wrapper around the `libbpf` library. It's a low-level library that allows you to load BPF programs into the kernel.
```

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+
+BPF programs can log messages to the kernel log. To enable logging, you need to set the license to `GPL`. Then you
+can use `bpf_printk` helper to log messages. The messages will be logged to the kernel log. You can get the messages
+ from `/sys/kernel/debug/tracing/trace_pipe`. Here is the best explanation that I found so far https://nakryiko.com/posts/bpf-tips-printk/
```
```suggestion
from `/sys/kernel/debug/tracing/trace_pipe`. Here is the best explanation that I found so far https://nakryiko.com/posts/bpf-tips-printk/.
```

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+
+Example:
+
+BPF: https://github.com/gravitational/teleport/blob/2fd8f75e38eebf5c6826ef594433e5165c3bfbe1/bpf/enhancedrecording/command.bpf.c#L18
+User space: https://github.com/gravitational/teleport/blob/2fd8f75e38eebf5c6826ef594433e5165c3bfbe1/lib/bpf/common_linux.go#L40-L51
```
```suggestion
* BPF: https://github.com/gravitational/teleport/blob/2fd8f75e38eebf5c6826ef594433e5165c3bfbe1/bpf/enhancedrecording/command.bpf.c#L18
* User space: https://github.com/gravitational/teleport/blob/2fd8f75e38eebf5c6826ef594433e5165c3bfbe1/lib/bpf/common_linux.go#L40-L51
```

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+The most important limitation is that you can't use any system calls in BPF programs. You can use only BPF helpers.
+The BPF programs are compiled using `clang` 10+, GCC is not supported. 
+
+User space code is located in `lib/bpf` directory. It's written in Go and uses `libbpfgo` library to load BPF programs into the kernel.
```
```suggestion
User space code is located in `lib/bpf` directory. It's written in Go and uses the `aquasecurity/libbpfgo` library to load BPF programs into the kernel.
```

Just so it's super clear which one. Various other suggestions follow this one.

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+User space code is located in `lib/bpf` directory. It's written in Go and uses `libbpfgo` library to load BPF programs into the kernel.
+`libbpfgo` is a Go wrapper around `libbpf` library. It's a low-level library that allows you to load BPF programs into the kernel.
+
+Note: `libbpfgo` is not backward compatible and you need a matching version of `libbpf` library. For that reason
+`libbpfgo` is not using semantic versioning, but tags have the following format `v0.4.5-libbpf-1.0.1`. 
+The first part is the version of `libbpfgo` and the second part is the version of `libbpf` library. 
+Using the wrong version of `libbpfgo` will result in a runtime/compilation error.
```
```suggestion
Note: `aquasecurity/libbpfgo` is not backwards compatible, you need to use the
exact `libbpf` version it requires. For that reason `aquasecurity/libbpfgo`
doesn't use semantic versioning, instead the tags have the format
`v0.4.5-libbpf-1.0.1`. The first part is the `aquasecurity/libbpfgo` version,
the second part is the `libbpf` version. Using the wrong version of `libbpf`
results in a runtime/compilation error.
```

Various minor changes, feel free to ignore.

codingllama commented on bpf/README.md:
```diff
@@ -0,0 +1,69 @@
+
+BPF programs can communicate with user space using maps. Maps are key-value stores that can be accessed from both
+BPF programs and user space. BPF programs can only access maps using BPF helpers. User space can access maps using
+`libbpfgo` library. Maps are defined in BPF programs and can be referenced by name from user space code.
```
```suggestion
`aquasecurity/libbpfgo` library. Maps are defined in BPF programs and can be referenced by name from user space code.
```

codingllama reviewed (commented):
Many thanks, very helpful!

jakule committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

jakule committed:
Update BPF README with additional details and instructions.

This commit expands on the instructions for setting up BPF, specifically clarifying that 'clang' can be installed through the 'apt install clang' command. A section detailing the requirements for libbpf installation has also been added for further clarity. Moreover, the BPF utilization in Teleport has been further explained, highlighting its usage for enhanced session recording and restricted networking. Lastly, the instructions for enabling LSM hooks on Ubuntu systems have been provided. This commit aims to provide comprehensive and updated information for configuring and understanding BPF usage.

jakule commented on bpf/README.md:
```diff
@@ -0,0 +1,95 @@
+It's implemented by using LSM hooks and BPF programs.
+https://goteleport.com/docs/server-access/guides/restricted-session/
+On ubuntu systems LSM hooks are not enabled in some version.
+Here are the instructions on how to enable them https://github.com/gravitational/teleport/issues/8089#issuecomment-924990678
```
I'll add this to our official docs.

codingllama commented on bpf/README.md:
```diff
@@ -68,4 +75,21 @@ BPF programs and user space. BPF programs can only access maps using BPF helpers
+* open family system calls
+* network connections
+
+All events are recorded in the audit log. https://goteleport.com/docs/server-access/guides/bpf-session-recording/
```
```suggestion
All events are recorded in the audit log. See https://goteleport.com/docs/server-access/guides/bpf-session-recording/.
```

codingllama commented on bpf/README.md:
```diff
@@ -68,4 +75,21 @@ BPF programs and user space. BPF programs can only access maps using BPF helpers
+All events are recorded in the audit log. https://goteleport.com/docs/server-access/guides/bpf-session-recording/
+
+Restricted networking allows you to restrict network access for users.
+It's implemented by using LSM hooks and BPF programs.
+https://goteleport.com/docs/server-access/guides/restricted-session/
```
```suggestion
It's implemented by using LSM hooks and BPF programs (
see https://goteleport.com/docs/server-access/guides/restricted-session/).
```

codingllama commented on bpf/README.md:
```diff
@@ -68,4 +75,21 @@ BPF programs and user space. BPF programs can only access maps using BPF helpers
+Restricted networking allows you to restrict network access for users.
+It's implemented by using LSM hooks and BPF programs.
+https://goteleport.com/docs/server-access/guides/restricted-session/
+On ubuntu systems LSM hooks are not enabled in some version.
```
```suggestion
On ubuntu systems LSM hooks are not enabled in some versions.
```

I think?

Could we be more specific? (Fine if not.)

jakule replied:
I think this is deterministic (each version that I personally checked had this disabled), but I think that would be a bold move to claim that LSM are always disabled based on that limited test.

codingllama commented on bpf/README.md:
```diff
@@ -68,4 +75,21 @@ BPF programs and user space. BPF programs can only access maps using BPF helpers
+It's implemented by using LSM hooks and BPF programs.
+https://goteleport.com/docs/server-access/guides/restricted-session/
+On ubuntu systems LSM hooks are not enabled in some version.
+Here are the instructions on how to enable them https://github.com/gravitational/teleport/issues/8089#issuecomment-924990678
```
```suggestion
Here are the instructions on how to enable them https://github.com/gravitational/teleport/issues/8089#issuecomment-924990678.
```

codingllama reviewed (commented):
Happy to take the PR as-is, if you want to move out of draft.

Many thanks!

jakule committed:
Apply suggestions from code review

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

jakule committed:
Merge branch 'master' into jakule/bpf-guide

jakule commented:
@codingllama Friendly ping

AntonAM commented on bpf/README.md:
```diff
@@ -0,0 +1,95 @@
+
+BPF programs are compiled into ELF files. ELF files have a license field that is used to verify that the BPF program
+is allowed to run in the kernel. The license field is set to `GPL` by default. Teleport uses `Dual BSD/GPL` which 
+disables some BPF features (like logging). To enable all BPF features to set the license to `GPL` in the BPF program
```
```suggestion
disables some BPF features (like logging). To enable all BPF features set the license to `GPL` in the BPF program
```

jakule committed:
Update bpf/README.md

Co-authored-by: Anton Miniailo <anton@goteleport.com>
